File: Coursera/Algorithms_Part_2/Ex23TopologicalSortestPath.java
/* *****************************************************************************
 *  Name:
 *  Date:
 *  Description:
 **************************************************************************** */

import edu.princeton.cs.algs4.In;
import edu.princeton.cs.algs4.Stack;
import edu.princeton.cs.algs4.StdOut;
import edu.princeton.cs.algs4.Stopwatch;

public class Ex23TopologicalSortestPath implements Ex21SortestPath {

    private Ex19DirectedEdge[] edgeTo;
    private double[] distTo;

    public Ex23TopologicalSortestPath(Ex20EdgeWeightedDigraph G, int s) {
        edgeTo = new Ex19DirectedEdge[G.V()];
        distTo = new double[G.V()];
        boolean[] marked = new boolean[G.V()];
        for (int i = 0; i < G.V(); i++) {
            distTo[i] = Double.POSITIVE_INFINITY;
            marked[i] = false;
        }

        distTo[s] = 0.0;
        Ex22TopologicalEWG topological = new Ex22TopologicalEWG(G);
        for (int i : topological.order()) {
            marked[i] = true;
            for (Ex19DirectedEdge e : G.adj(i)) {
                if (!marked[e.to()]) relax(e);
            }
        }
    }

    @Override
    public double distTo(int v) {
        return distTo[v];
    }

    @Override
    public Iterable<Ex19DirectedEdge> pathTo(int v) {
        Stack<Ex19DirectedEdge> path = new Stack<>();
        for (Ex19DirectedEdge e = edgeTo[v]; e != null; e = edgeTo[e.from()])
            path.push(e);
        return path;
    }

    @Override
    public boolean hasPathTo(int v) {
        return edgeTo[v] != null;
    }

    private void relax(Ex19DirectedEdge e) {
        int v = e.from(), w = e.to();
        if (distTo[w] > distTo[v] + e.weight()) {
            distTo[w] = distTo[v] + e.weight();
            edgeTo[w] = e;
        }
    }

    public static void main(String[] args) {
        Stopwatch stopwatch = new Stopwatch();
        In in = new In("1000EWD.txt");
        Ex20EdgeWeightedDigraph G = new Ex20EdgeWeightedDigraph(in);
        Ex23TopologicalSortestPath sp = new Ex23TopologicalSortestPath(G, 0);
        StdOut.println(G);

        for (int i = 1; i < G.V(); i++) {
            StdOut.printf("distTo " + i + ": %.2f path: ", sp.distTo(i));
            for (Ex19DirectedEdge e : sp.pathTo(i))
                StdOut.print(e + " ");
            StdOut.println();
        }

        StdOut.println();
        StdOut.println("Time Elapsed: " + stopwatch.elapsedTime());
    }
}

File: PythonLearnig/Algorithms_Part_2/Ex23TopologicalSortestPath.py
from Ex20EdgeWeightedDigraph import EdgeWeightedDigraph
from Ex19DirectedEdge import DirectedEdge
from Ex22TopologicalEWD import TopologicalEWD


class TopologicalSortestPath:

    def __init__(self, digraph, source):
        self.edge_to = [None] * digraph.get_v()
        self.dist_to = [float('inf')] * digraph.get_v()

        marked = [False] * digraph.get_v()

        self.dist_to[source] = 0.0
        topological = TopologicalEWD(digraph)
        for v in topological.order():
            marked[v] = True
            for edge in digraph.get_adj(v):
                if not marked[edge.to_v()]:
                    self.relax(edge)

    def path_to(self, v):
        if self.edge_to[v] is not None:
            for path in self.path_to(self.edge_to[v].from_v()):
                yield path
            yield self.edge_to[v]

    def has_path_to(self, v):
        return self.edge_to[v] is not None

    def relax(self, edge):
        v = edge.from_v()
        w = edge.to_v()
        if self.dist_to[v] + edge.weight < self.dist_to[w]:
            self.dist_to[w] = self.dist_to[v] + edge.weight
            self.edge_to[w] = edge


def test_topological_sortest_path():

    digraph = EdgeWeightedDigraph(0)
    digraph.init_from('tinyEWD.txt')
    print()
    print(digraph)

    sp = TopologicalSortestPath(digraph, 0)
    for v in range(digraph.get_v()):
        builder = f'Has path to {v}? {sp.has_path_to(v)}'
        if sp.has_path_to(v):
            builder += f' dist: {round(sp.dist_to[v], 2)} path:'
            for path in sp.path_to(v):
                builder += f'\n\t{path}'
        print(builder)


if __name__ == '__main__':
    test_topological_sortest_path()
